[FLINK-13550][rest][ui] Vertex Flame Graph (#15054)
apache/flink: docs/content/docs/ops/debugging/flame_graphs.md

---
title: "Flame Graphs"
weight: 3
type: docs
aliases:
  - /ops/debugging/flame_graphs.html
  - /ops/debugging/flame_graphs
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Flame Graphs

> [Flame Graphs](http://www.brendangregg.com/flamegraphs.html) are a visualization that effectively surfaces answers to the questions like:

Reviewer: ```suggestion
[Flame Graphs](http://www.brendangregg.com/flamegraphs.html) are a visualization that effectively surfaces answers to questions like:
```

- Which methods are currently consuming CPU resources?

> - How consumption by one method compares to the others?

Reviewer: ```suggestion
- How does consumption by one method compare to the others?
```

{{< img src="/fig/flame_graph_operator.png" class="img-fluid" width="90%" >}}
{{% center %}}
Operator's On-CPU Flame Graph
{{% /center %}}

Apart from the On-CPU Flame Graphs, [Off-CPU](http://www.brendangregg.com/FlameGraphs/offcpuflamegraphs.html) and Mixed visualizations are available and can be switched between by using the selector at the top of the pane:

{{< img src="/fig/flame_graph_selector.png" class="img-fluid" width="30%" >}}

The Off-CPU Flame Graph visualizes blocking calls found in the samples. A distinction is made as follows:
- On-CPU: `Thread.State` in **[RUNNABLE, NEW]**
- Off-CPU: `Thread.State` in **[TIMED_WAITING, WAITING, BLOCKED]**

{{< img src="/fig/flame_graph_off_cpu.png" class="img-fluid" width="90%" >}}
{{% center %}}
Off-CPU Flame Graph
{{% /center %}}

Mixed mode Flame Graphs are constructed from stack traces of threads in all possible states.

{{< img src="/fig/flame_graph_mixed.png" class="img-fluid" width="90%" >}}
{{% center %}}
Flame Graph in Mixed Mode
{{% /center %}}

##  Sampling process

> Stack traces collection happens purely within a JVM, so only method calls within the Java runtime are visible (no system calls).

Reviewer: ```suggestion
The collection of stack traces is done purely within the JVM, so only method calls within the Java runtime are visible (no system calls).
```

Flame Graph construction is performed at the level of an individual [operator]({{< ref "docs/concepts/glossary" >}}#operator), i.e. all [task]({{< ref "docs/concepts/glossary" >}}#task) threads of that operator are sampled in parallel and their stack traces are combined. 

{{< hint info >}}
**Note:** 

> Stack trace samples from all threads of an operator are combined together. If a method call consumes 100%  of resources in one of the parallel tasks but none in the others, the bottleneck might be obcured by averaging out.   

Reviewer: ```suggestion
Stack trace samples from all threads of an operator are combined together. If a method call consumes 100% of the resources in one of the parallel tasks but none in the others, the bottleneck might be obscured by being averaged out.   
```
